clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -2867,6 +2913,163 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::next_record(output_ordinal_t outp
+                    // branch target we're looking for. This could be the next user-space
+                    // instruction after the kernel sequence ends, or for
+                    // syscall-to-switch it may be the first instruction of the switch
+                    // sequence.
+                    if (outputs_[output].kibt_readahead_q.instrs_count() == 2 &&
+                        record_type_is_instr(next_record, &maybe_next_pc)) {
+                        next_pc = maybe_next_pc;
+                    } else if (record_type_is_marker(next_record, marker_type,
+                                                     marker_value) &&
+                               marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT) {
+                        // If there was a signal after the injected kernel trace, we
+                        // indicate that the kernel-trace-final indirect branch is to
+                        // the interrupted pc.
+                        next_pc = marker_value;
+                    } else if (record_type_is_thread_exit(next_record) &&
+                               // For thread-sharded parallel analysis, we don't
+                               // want a thread-final syscall/switch branch to point
+                               // to the next thread in the same output stream.
+                               options_.mapping ==
+                                   sched_type_t::MAP_TO_CONSISTENT_OUTPUT) {
+                        // We let next_pc remain zero, to indicate that the indirect
+                        // branch instruction should be removed.
+                        break;
+                    }
+                }
+            }
+            if (next_pc == 0) {
+                // We couldn't find the target of the indirect branch instruction. Since
+                // we do not want to end the trace at a branch with an unknown target pc,
+                // we remove the indirect branch (same as how we do for user-space-only
+                // traces). Kernel sequences do not contain consecutive returns.
+                int discard_count = 0;
+                do {
+                    ++discard_count;
+                    outputs_[output].kibt_readahead_q.pop();
+                    assert(!outputs_[output].kibt_readahead_q.empty());
+                    to_return = outputs_[output].kibt_readahead_q.front();
+                    is_marker = record_type_is_marker(to_return.record, marker_type,
+                                                      marker_value);
+                } while (!(is_marker &&
+                           (marker_type == TRACE_MARKER_TYPE_SYSCALL_TRACE_END ||
+                            marker_type == TRACE_MARKER_TYPE_CONTEXT_SWITCH_END)));
+                VPRINT(this, 2,
+                       "next_record[%d]: Did not find a target pc for %s, removing "
+                       "indirect branch records from trace end; discard %d records\n",
```

Comment:
Why "records"? Isn't there only one? You said above they do not contain consecutive returns.

Reply:
"records" (plural) because for record_scheduler_t which looks at trace_entry_t, it would mean two records (the branch_target marker and the indirect branch instr).